### PR TITLE
Fix issue #6944: Optimize response/request_body_buf by using list of byte chunks to avoid concatenation overhead

### PR DESCRIPTION
**Pull request by jackfromeast**
#### Description

Fix issue #6944: Optimize response/request_body_buf by using a list of byte chunks to avoid concatenation overhead. 

#### Checklist

 - [x] I have updated tests where applicable.
 - [x] I have added an entry to the CHANGELOG.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,8 @@
   ([#6921](https://github.com/mitmproxy/mitmproxy/pull/6921), @zendai)
 * Add `HttpConnectedHook` and `HttpConnectErrorHook`
   ([#6930](https://github.com/mitmproxy/mitmproxy/pull/6930), @errorxyz)
+* Fix the issue of non-linear growth in processing time in mitmproxy related to packet size.
+  ([#6944](https://github.com/mitmproxy/mitmproxy/issues/6944), @jackfromeast)
```
Comment (errorxyz) [inline CHANGELOG.md:15]: This line should be added automatically by the CI, you should remove this line

File: mitmproxy/proxy/layers/http/__init__.py [hunk 2]
```diff
@@ -145,8 +145,10 @@ class DropStream(HttpCommand):
 
 
 class HttpStream(layer.Layer):
-    request_body_buf: bytes
-    response_body_buf: bytes
+    request_body_buf: list
+    response_body_buf: list
```
Comment (errorxyz) [inline mitmproxy/proxy/layers/http/__init__.py:149]: I think a `bytearray` type might be more appropriate for this purpose

Comment (mhils) [reply mitmproxy/proxy/layers/http/__init__.py:149]: I _think_ `list[bytes]` might be better here: Appending is amortized O(1), so N slices can be appended in O(n). Joining is O(n) as well: https://github.com/python/cpython/blob/65a12c559cbc13c2c5a4aa65c76310bd8d2051a7/Objects/stringlib/join.h#L57-L130. 

For bytearray I am not so sure. Is there anything in the Python docs that talks about how it does memory allocation?

Comment (mhils) [reply mitmproxy/proxy/layers/http/__init__.py:149]:
```
class ReceiveBuffer:
    _chunks: list[bytes]
    _len: int

    def __init__(self):
        self._chunks = []
        self._len = 0

    def __iadd__(self, other: bytes):
        assert isinstance(other, bytes)
        self._chunks.append(other)
        self._len += len(other)
        return self

    def __len__(self):
        return self._len

    def __bytes__(self):
        return b"".join(self._chunks)

class ReceiveBufferWithByteArray:
    _buf: bytearray

    def __init__(self):
        self._buf = bytearray()

    def __iadd__(self, other: bytes):
        self._buf.extend(other)
        return self

    def __len__(self):
        return len(self._buf)

    def __bytes__(self):
        return bytes(self._buf)


if __name__ == "__main__":
    import time
    chunk_size = 65_000  # ~ one TCP packet
    num_chunks = 1024 * 1024 * 100 // chunk_size  # 100mb

    data_chunk = b'a' * chunk_size

    def bench(buf: ReceiveBuffer | ReceiveBufferWithByteArray):
        start = time.time()

        for _ in range(num_chunks):
            buf += data_chunk
        ret = bytes(buf)

        print(buf.__class__.__name__, time.time() - start)
        return ret

    _ = bench(ReceiveBuffer())
    _ = bench(ReceiveBufferWithByteArray())
```

```
ReceiveBuffer 0.021017074584960938
ReceiveBufferWithByteArray 0.2989828586578369
```

:-)

**Pull request merged**